templates/terraform/processing-project/buckets.tf

```diff
@@ -24,20 +24,32 @@ resource google_storage_bucket temp_bucket {
 }
 
 # Bucket for temporary data storage.
-# TODO: Add an IAM resource which allows the repo to read from this.
 resource google_storage_bucket staging_storage {
   provider = google.target
   name = "${var.project_name}-staging-storage"
   location = "US"
 
-  lifecycle_rule {
-    action {
-      type = "Delete"
-    }
-
-    # Delete files after they've been in the bucket for 30 days.
-    condition {
-      age = 30
+  dynamic "lifecycle_rule" {
+    for_each = var.deletion_age == null ? [] : [var.deletion_age]
+    content {
+      action {
+        type = "Delete"
+      }
+      condition {
+        age = lifecycle_rule.value
+      }
     }
   }
 }
+
+resource google_storage_bucket_iam_member test_iam {
+  # for_each doesn't like lists, so we convert it to a set
+  for_each = toset(var.access_emails)
+
+  provider = google.target
+  bucket = google_storage_bucket.staging_storage.name
+  # The legacyBucketReader policy seems to be the correct role to have the equivalent of a "Bucket Reader" ACL,
+  # as noted here https://cloud.google.com/storage/docs/access-control/iam
+  role = "roles/storage.legacyBucketReader"
+  member = "user:${each.value}"
```

Comment:
Ergh, I don't think we can hard-code `user` here: https://www.terraform.io/docs/providers/google/r/storage_bucket_iam.html#member-members

There are a few things we could do:
1. (Most flexible) Change `access_emails` to a `list(object({type = string, email = string}))`
2. (Most opintionated) Change `access_emails` to something like `reader_groups`, and hard-code `group` here

Option 2 follows best-practices, but I'm not sure if we can get away with it. We could always add more specific names for any service accounts that need to get access to the bucket (like the Jade SA).

Option 1 is annoying  because it's more verbose than just writing out the `<type>:<email>` strings, but it does give a bit more structure than just a freetext string. It could be slightly better if we bump to TF 0.12.20 and use its new support for custom input validation: https://www.terraform.io/docs/configuration/variables.html#custom-validation-rules

Reply:
Ah I didn't realize we'd want to use this list to add non-user emails as well. Using groups is probably best but yeah I figured it'd be easier/quicker to get working with just straight emails. Option 1 isn't that much more code! I'm in favor of that implementation for this/for now

Reply:
So a `list(object({type = string, email = string}))` doesn't quite work with this `for_each` pattern; how averse are we to having the input string there be the entire "<type>:<address>"? I can try to bump versions for the input validation. At this point I'm also more open to going the group route lol

Reply:
I'm +1 to forcing groups, -1 to an input that requires knowing TF's magic strings.

You could also try using `count` instead of `for_each`: https://www.terraform.io/docs/configuration/resources.html#count-multiple-resource-instances-by-count

Reply:
For posterity: The decision is forcing groups

Reply:
+1 for showing me a thing i didn't know possible in tf a for loop inside a conditional operator.